hammer/flowgraph/flowgraph.py

```diff
@@ -0,0 +1,362 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+#  Represents a flowgraph in HAMMER that can be run and verified.
+#  See README.config for more details.
+#
+#  See LICENSE for licence details.
+
+# pylint: disable=invalid-name
+
+import json
+import os
+import uuid
+from dataclasses import dataclass, field, asdict
+from enum import Enum
+from typing import Any, Union
+
+import networkx as nx
+from networkx.readwrite import json_graph
+
+from hammer.logging import HammerVLSILogging
+from hammer.vlsi.cli_driver import CLIDriver
+
+
+class Status(Enum):
+    """Represents the status of a node in the flowgraph.
+    """
+    NOT_RUN    = "NOT_RUN"
+    RUNNING    = "RUNNING"
+    INCOMPLETE = "INCOMPLETE"
+    INVALID    = "INVALID"
+    COMPLETE   = "COMPLETE"
+
+
+@dataclass
+class Node:
+    """Defines a node for an action in a flowgraph.
+
+    Returns:
+        Node: Complete description of an action.
+    """
+    action:            str
+    tool:              str
+    pull_dir:          str
+    push_dir:          str
+    required_inputs:   list[str]
+    required_outputs:  list[str]
+    status:            Status    = Status.NOT_RUN
+    # __uuid:            uuid.UUID = field(default_factory=uuid.uuid4)
+    driver:            CLIDriver = field(default_factory=CLIDriver)
+    optional_inputs:   list[str] = field(default_factory=list)
+    optional_outputs:  list[str] = field(default_factory=list)
+    step_controls:     dict[str, str] = field(default_factory=lambda: {
```

Comment:
Techincally, these flow control flags are legacy. Recommended ones are now `--start_before_step`, `--start_after_step`, etc. so need to add those into the dict.